Improve documentation & test coverage (#29)
slashmo/gsoc-swift-baggage-context: README.md

]
```

## Usage

`BaggageContext` is intended to be used in conjunction with the instrumentation of distributed systems. To make this
instrumentation work, all parties involved operate on the same `BaggageContext` type. These are the three common
parties, in no specific order, and guidance on how to use `BaggageContext`:

> ### Frameworks/Libraries

Reviewer: ```suggestion
### Library & Framework Authors - passing context and instrumenting libraries
```

Developers creating frameworks/libraries (e.g. NIO, gRPC, AsyncHTTPClient, ...) which benefit from being instrumented
should adopt `BaggageContext` as part of their public API. AsyncHTTPClient for example might accept a context like this:

```swift
let context = BaggageContext()
client.get(url: "https://swift.org", context: context)
```

For more information on where to place this argument and how to name it, take a look at the

> ["Context-Passing Guidelines"](#Context-Passing-Guidelines).

Reviewer: ```suggestion
[Context-Passing Guidelines](#Context-Passing-Guidelines).
```

> Generally speaking, framework developers do not care about the specific contents of the `BaggageContext` but need to

Reviewer: ```suggestion
Generally speaking, frameworks and libraries should treat baggage as an _opaque container_ and simply thread it along all asynchronous boundaries a call may have to go through. Libraries and frameworks should not attempt to reuse context as a means of passing values that they need for "normal" operation.
```

> make sure to feed it through the entire system. At cross-cutting boundaries, e.g. right before sending an HTTP

Reviewer: ```suggestion
At cross-cutting boundaries, e.g. right before sending an HTTP
```

> request, they'd inject the `BaggageContext` into the HTTP headers, allowing context propagation. On the flip-sight, an

Reviewer: ```suggestion
request, they inject the `BaggageContext` into the HTTP headers, allowing context propagation. On the receiving side, an
```

less colloquial

> request, they'd inject the `BaggageContext` into the HTTP headers, allowing context propagation. On the flip-sight, an

Reviewer: ```suggestion
request, they'd inject the `BaggageContext` into the HTTP headers, allowing context propagation. On the receiving side, an
```

(was typo anyway with "sight" 😉 )

Author: Added through https://github.com/slashmo/gsoc-swift-baggage-context/pull/29/files#r478818645

HTTP server should extract the request headers into a `BaggageContext`. Injecting/extracting is part of the
`swift-tracing` libraries [and documented in its own repository](https://github.com/slashmo/gsoc-swift-tracing).

> ### Instrumentation

Reviewer: ```suggestion
### Instrumentation Authors - defining, injecting and extracting baggage
```

When implementing instrumentation for cross-cutting tools, `BaggageContext` becomes the way you propagate metadata such
as trace ids. Because each instrument knows what values might be added to the `BaggageContext` they are the ones
creating `BaggageContextKey` types dictating the type of value associated with each key added to the context:

```swift
private enum TraceIDKey: BaggageContextKey {
  typealias Value = String

> }

Reviewer: Please do take the extra step to always define:

```swift
extension BaggageContextProtocol {
    var traceID: TraceIDKey.Value? {
        get {
            return self[TraceIDKey.self]
        }
        set {
            self[TraceIDKey.self] = newValue
        }
    }
```

Reviewer: The keys are an implementation detail

Author: ```suggestion
}

extension BaggageContextProtocol {
    var traceID: TraceIDKey.Value? {
        get {
            return self[TraceIDKey.self]
        }
        set {
            self[TraceIDKey.self] = newValue
        }
    }
}
```

Reviewer: Yeap, please apply

var context = BaggageContext()
context[TraceIDKey.self] = "4bf92f3577b34da6a3ce929d0e0e4736"
print(context[TraceIDKey.self] ?? "new trace id")
```

**TODO**: Add paragraph on why keys should be private/internal (#25)

> To make accessing values a bit more convenient, we encourage you to add computed properties to `BaggageContextProtocol`:

Reviewer: I see, you added it below. I'd fold it all into 1 snippet; It's not optional, noone should be using the Bla.self style but always though these accessors.

We can also consider if what we do with attributes in tracing can work here or not 🤔  New ticket perhaps.

```swift
extension BaggageContextProtocol {
  var traceID: String? {
    get {
      return self[TraceIDKey.self]
    }
    set {
      self[TraceIDKey.self] = newValue
    }
  }
}
```

> ### Distributed System

Reviewer: I'd rename to 

```suggestion
### End Users - explicit context passing
```

or something like that; It is crucial to not scare away people from this concept. People may not be really fully aware they're building distributed systems ("but i just call the database" etc). We should absolutely normalize passing around contexts, it is not weird or special, it's just status quo of today's server development ecosystem.

Reviewer: Please make this sub-section the _first_ of the three.

We should go in increasing order of complexity:

- end users
- library authors
- instrumentation authors

Author: Yeah, that's a good point. I agree that distributed systems might be a bit intimidating. The reordering also makes sense to me as that's probably also the decreasing order of the amounts of developers in these fields.

When creating distributed systems, each of your components will likely make use of one or more generically instrumented
frameworks/libraries. This means you get to choose what instrument(s) to use 🙌. Check out

> [the swift-tracing](https://github.com/slashmo/gsoc-swift-tracing) for instructions on how to get up & running.

Reviewer: I'd not prefix this with "when creating distributed systems ..." at all, but simply say "you see some API that takes context, you pass whatever context you already have"

You can get inspiration on wording here: https://golang.org/pkg/context/

We should really focus on "you see a context? you pass one" if you don't have one, stop and think why, if you really dont have one, pass an empty one -- this relates to: https://github.com/slashmo/gsoc-swift-baggage-context/issues/26 as well

Reviewer: Then after we explain "just always pass the context"

we can add a section on "This will enable you to enable tracing whenever your application grows", there will be no "oh no, now we have to fix all places to pass context" -- once you _suddenly_ need tracing because there's a fire on production or you're profiling some troublesome performance regressions, no-one will have the time to go through the entire system and start passing contexts around. You should always pass it, so that you're ready for when you need it -- your future self will be thankful for it 👍 

Or something like that

Occasionally, when using multiple frameworks/libraries together, you'll need to pass along a `BaggageContext` to ensure

> that the values stored in this context get propagated correctly.

Reviewer: I think we can remove this sentence; passing context is simply everywhere, and we have some convenience for the many libraries case where all have their own context things.

Author: 👍

## Context-Passing Guidelines

For context-passing to feel consistent and Swifty among all server-side (and not only) libraries and frameworks
aiming to adopt `BaggageContext` (or any of its uses, such as Distributed Tracing), we suggest the following set of
guidelines:

### Argument naming/positioning

> To propagate baggage through function calls (and asynchronous-boundaries) it may often be necessary to pass it

Reviewer: ```suggestion
Propagating baggage context through your system is to be done explicitly, meaning as a parameter in function calls, following the "flow" of execution. 

> Do not store the context object in your classes or structs as this will inevitably lead to mistakes and hard to track bugs.
```

  2. Defaulted non-function parameters (e.g. `(mode: Mode = .default)`),
  3. Required function parameters, including required trailing closures (e.g. `(onNext elementHandler: (Value) -> ())`),
  4. Defaulted function parameters, including optional trailing closures (e.g. `(onComplete completionHandler: (Reason) -> ()) = { _ in }`).

> - Baggage Context should be passed as **the last parameter in the required non-function parameters group in a function declaration**.

Reviewer: 👍

In case there are _multiple_ "framework-ish" parameters, such as passing a NIO `EventLoop` or similar, we suggest:

- `func perform(_ work: Work, for user: User,` _`frameworkThing: Thing, eventLoop: NIO.EventLoop,`_ **`context: BaggageContext`** `)`
  - pass the baggage as **last** of such non-domain specific parameters as it will be _by far more_ omnipresent than any
  specific framework parameter - as it is expected that any framework should be accepting a context if it can do so.
  While not all libraries are necessarily going to be implemented using the same frameworks.

> We feel it is important to preserve Swift's human-readable nature of function definitions. In other words, we intend to

Reviewer: 👍
